flow/display_list_canvas.cc

```diff
@@ -170,8 +171,13 @@ void DisplayListCanvasDispatcher::drawAtlas(const sk_sp<SkImage> atlas,
 void DisplayListCanvasDispatcher::drawPicture(const sk_sp<SkPicture> picture,
                                               const SkMatrix* matrix,
                                               bool render_with_attributes) {
-  canvas_->drawPicture(picture, matrix,
-                       render_with_attributes ? &paint() : nullptr);
+  if (render_with_attributes) {
+    // drawPicture does an implicit saveLayer if an SkPaint is supplied.
+    TRACE_EVENT0("flutter", "Canvas::saveLayer");
+    canvas_->drawPicture(picture, matrix, &paint());
+  } else {
+    canvas_->drawPicture(picture, matrix, nullptr);
+  }
```

Comment:
We may also want to capture when drawPaint and draw color will insert a save layer.

Reply:
It looks like there is a public way to determine if the drawPaint will use a saveLayer so I'm adding that. (It's only when there is an ImageFilter, though, so it probably won't happen much in practice.)

drawColor doesn't show any signs of using a saveLayer, though. The operation becomes a drawPaint with no ImageFilter which never generates a saveLayer in the Skia code.

Reply:
I should probably also mark it as recorded when I record the drawPaint. Should that be the same event name as when we record saveLayer? Or modified slightly - like "(drawPaint Recorded)" to distinguish it so that developers can know to look for a drawPaint call instead of a saveLayer. Or "ui.Canvas::drawPaint (Recorded needing saveLayer)" or something?

Reply:
Hmm. I know the method name has something about ImageFilters in its return type, but when I was tracing through the code it looked like the surface could decide to trigger a savelayer with or without an image filter... Specifically I'm looking at these sites:

https://github.com/google/skia/blob/6a5f772ea6f0f689b32e9dbdb78f6a24963f524a/src/core/SkCanvas.cpp#L1750 eventually ends up in https://github.com/google/skia/blob/6a5f772ea6f0f689b32e9dbdb78f6a24963f524a/src/core/SkCanvas.cpp#L170 and then https://github.com/google/skia/blob/6a5f772ea6f0f689b32e9dbdb78f6a24963f524a/src/core/SkCanvas.cpp#L170

What's the part that I'm missing?

Reply:
Sorry, tha tlast link was supposed to be https://github.com/google/skia/blob/main/src/image/SkSurface.cpp#L120

But now I see the part you linked and it all makes sense. Ok.

Reply:
Yeah, I think that call can reject the operation entirely, but the actual saveLayer happens in their ...Auto... object.